Zend/zend_object_handlers.c

```diff
@@ -1334,15 +1332,29 @@ ZEND_API void zend_std_unset_property(zend_object *zobj, zend_string *name, void
 	if (EXPECTED(IS_VALID_PROPERTY_OFFSET(property_offset))) {
 		zval *slot = OBJ_PROP(zobj, property_offset);
 
-		if (Z_TYPE_P(slot) != IS_UNDEF) {
-			if (UNEXPECTED(prop_info && (prop_info->flags & ZEND_ACC_READONLY))) {
-				if (Z_PROP_FLAG_P(slot) & IS_PROP_REINITABLE) {
-					Z_PROP_FLAG_P(slot) &= ~IS_PROP_REINITABLE;
-				} else {
-					zend_readonly_property_unset_error(prop_info->ce, name);
-					return;
+		if (prop_info && UNEXPECTED(prop_info->flags & (ZEND_ACC_READONLY|ZEND_ACC_PPP_SET_MASK))) {
+			if (Z_TYPE_P(slot) == IS_UNDEF
+			 && !(Z_PROP_FLAG_P(slot) & IS_PROP_UNINIT)
+			 && zobj->ce->__unset) {
+				goto magic_unset;
+			}
+			if (prop_info->flags & ZEND_ACC_READONLY) {
+				if (Z_TYPE_P(slot) != IS_UNDEF) {
+					if (Z_PROP_FLAG_P(slot) & IS_PROP_REINITABLE) {
+						Z_PROP_FLAG_P(slot) &= ~IS_PROP_REINITABLE;
```

Comment:
I removed this branch entirely, because it doesn't make much sense. We don't need to remove this flag, because unsetting the property will make sure that it will be writable anyway. So, we just skip the error now if the property has this flag.